packages/std/src/mock.rs

```diff
@@ -129,26 +129,28 @@ pub fn mock_env<T: Api, U: Into<HumanAddr>>(api: &T, sender: U, sent: &[Coin]) -
 
 /// MockQuerier holds an immutable table of bank balances
 /// TODO: also allow querying contracts
-pub struct MockQuerier<C: Clone + DeserializeOwned = Never> {
+pub struct MockQuerier<C: DeserializeOwned = Never> {
     bank: BankQuerier,
     staking: StakingQuerier,
     // placeholder to add support later
     wasm: NoWasmQuerier,
-    // TODO: Replace argument C -> &C. This requires proper lifetime handling ...
-    handle_custom: Box<dyn Fn(C) -> QuerierResult>, // use box to avoid the need of another generic type
+    handle_custom: Box<dyn for<'a> Fn(&'a C) -> QuerierResult>, // use box to avoid the need of another generic type
 }
 
-impl<C: Clone + DeserializeOwned> MockQuerier<C> {
+impl<C: DeserializeOwned> MockQuerier<C> {
     pub fn new(balances: &[(&HumanAddr, &[Coin])]) -> Self {
+        // strange argument notation suggested as a workaround here: https://github.com/rust-lang/rust/issues/41078#issuecomment-294296365
+        let no_handler = |_: &_| -> QuerierResult {
+            Err(SystemError::UnsupportedRequest {
+                kind: "custom".to_string(),
+            })
+        };
```

Comment:
Can't this be a free floating function?:
```rust
fn no_handler<C: DeserializeOwned>(_: &C) -> QuerierResult {
    Err(/* All the rest */)
}
```

Reply:
I think so. In the contract there is `fn execute`, which demonstrates what you suggest. Is there a good reason for one way or the other?

Reply:
They're equivalent, really, except a free floating function is reusable. But maybe you on't want it to be reusable. More than anything i didn't like the comment :sweat_smile: I think you don't need the `: &_` part

Reply:
Yeah, don't like the comment as well. But `|_|` did not work 🤷‍♂️

But I think storing an anoumous function to a variable is pointless and it should be an 'fn' instead, no matter where it lives.

Reply:
Using a free function creates lifetime issues again :(

```
error[E0310]: the parameter type `C` may not live long enough
   --> packages/std/src/mock.rs:160:29
    |
148 | impl<C: DeserializeOwned> MockQuerier<C> {
    |      -- help: consider adding an explicit lifetime bound...: `C: 'static +`
...
160 |             custom_handler: Box::from(no_handler),
    |                             ^^^^^^^^^^^^^^^^^^^^^
    |
note: ...so that the type `for<'r> fn(&'r C) -> std::result::Result<std::result::Result<encoding::Binary, errors::std_error::StdError>, errors::system_error::SystemError> {mock::MockQuerier::<C>::new::no_handler::<C>}` will meet its required lifetime bounds
   --> packages/std/src/mock.rs:160:29
    |
160 |             custom_handler: Box::from(no_handler),
    |                             ^^^^^^^^
```

Reply:
and in an `fn` I cannot use `&_`, which seems to work in closures

Reply:
related to the large comment below, Box requires that it's contents are `'static`, so `no_handler` must be `'static`, but it contains a generic type `C` which was not constrained as `'static`

Reply:
Okay, thank you. I get the idea. Turns out 

```rust
fn no_handler<C: DeserializeOwned + 'static>(_: &C) -> MockQuerierCustomHandlerResult {
    Err(SystemError::UnsupportedRequest {
        kind: "custom".to_string(),
    })
}
```

does the job. Unfortunately this `+ 'static` needs to be added in all of the stack there `C: DeserializeOwned` is used now.

So as long as there is no very good reason for the free function, I use the closure with the strange signature and the comment. But I inline it instead of assigning it to a variable.

Reply:
cool, i'll resolve this then